File: Algorithms and Data Structure/Sorting/MergeSort.cpp
void merge(int arr[], int left, int middle, int right){

    int sizeOfLeftPart = middle-left+1;
    int sizeOfRightPart = right-middle;
    int leftPart[sizeOfLeftPart];
    int rightPart[sizeOfRightPart];

    for (int i = 0; i < sizeOfLeftPart; i++)
        leftPart[i] = arr[left + i];
    for (int i = 0; i < sizeOfRightPart; i++)
        rightPart[i] = arr[middle + 1 + i];

    int i = 0, j = 0, k = left;
    while(i < sizeOfLeftPart && j < sizeOfRightPart){
        if (leftPart[i] <= rightPart[j]){
            arr[k] = leftPart[i];
            i++;
        }else{
            arr[k] = rightPart[j];
            j++;
        }
        k++;
    }

    while (i < sizeOfLeftPart){
        arr[k] = leftPart[i];
        i++;
        k++;
    }

    while (j < sizeOfRightPart){
        arr[k] = rightPart[j];
        j++;
        k++;
    }
}

void MergeSort(int arr[], int l, int r){
    if (l < r){

        int middle = l+(r-l)/2;

        mergeSort(arr, l, middle);
        mergeSort(arr, middle+1, r);

        merge(arr, l, middle, r);
    }
}

File: Algorithms and Data Structure/go/algorithms/sorting/merge-sort/mergesort.go
package merge

func sort(arr []int) {
	if 0 < len(arr) {
		middle := len(arr) / 2
		sort(arr[:middle])
		sort(arr[middle:])

		merge(arr, middle)
	}
}

func merge(arr []int, middle int) {
	leftPart := arr[:middle]
	rightPart := arr[middle:]

	var i int
	var j int
	var k int

	for i < len(leftPart) && j < len(rightPart) {
		if leftPart[i] < rightPart[j] {
			arr[k] = leftPart[i]
			i++
		} else {
			arr[k] = rightPart[j]
			j++
		}
		k++
	}

	for i < len(leftPart) {
		arr[k] = leftPart[i]
		i++
		k++
	}

	for j < len(rightPart) {
		arr[k] = rightPart[i]
		j++
		k++
	}
}
